File: 287_find_the_duplicate_number.rb
=begin
Given an array nums containing n + 1 integers where each integer is between 1 and n (inclusive), prove that at least one duplicate number must exist. Assume that there is only one duplicate number, find the duplicate one.

Example 1:

Input: [1,3,4,2,2]
Output: 2
Example 2:

Input: [3,1,3,4,2]
Output: 3
Note:

You must not modify the array (assume the array is read only).
You must use only constant, O(1) extra space.
Your runtime complexity should be less than O(n2).
There is only one duplicate number in the array, but it could be repeated more than once.
=end

# @param {Integer[]} nums
# @return {Integer}
def find_duplicate(nums)
    slow = nums[0]
    fast = nums[0]
    loop do
        slow = nums[slow]
        fast = nums[nums[fast]]
        
        if slow == fast
            break
        end
    end

    p1 = nums[0]
    p2 = slow
    while p1 != p2
        p1 = nums[p1]
        p2 = nums[p2]
    end
    
    return p1
end

File: problems/287_find_the_duplicate_number.rs
/*
Given an array of integers nums containing n + 1 integers where each integer is in the range [1, n] inclusive.

There is only one repeated number in nums, return this repeated number.

You must solve the problem without modifying the array nums and uses only constant extra space.

 

Example 1:

Input: nums = [1,3,4,2,2]
Output: 2
Example 2:

Input: nums = [3,1,3,4,2]
Output: 3
 

Constraints:

1 <= n <= 105
nums.length == n + 1
1 <= nums[i] <= n
All the integers in nums appear only once except for precisely one integer which appears two or more times.
 

Follow up:

How can we prove that at least one duplicate number must exist in nums?
Can you solve the problem in linear runtime complexity?
*/

impl Solution {
    pub fn find_duplicate(nums: Vec<i32>) -> i32 {
        let mut p1 = 0;
        let mut p2 = 0;

        loop {
            p1 = nums[p1 as usize];
            p2 = nums[nums[p2 as usize] as usize];
            if p1 == p2 {
                break;
            }
        }

        p1 = 0;
        while p1 != p2 {
            p1 = nums[p1 as usize];
            p2 = nums[p2 as usize];
        }

        p1
    }
}

